### PR TITLE
Fixing 2947: Call to ParkingWindow.RemoveReflectChild() causes Cross-thread exception under load

### LINKED ISSUE #2947: Call to ParkingWindow.RemoveReflectChild() causes Cross-thread exception under load
* .NET Core Version: 

Version: 5.0.0-alpha.1.20072.3

* Have you experienced this same bug with .NET Framework?: 
No

**Problem description:**

Call to ParkingWindow.RemoveReflectChild() causes Cross-thread exception under load

**Expected behavior:**
There should be no error when GC frees the resources.

**Minimal repro:**
Run WinForms app under load and trigger GC to free resources.

```
System.InvalidOperationException: 'Cross-thread operation not valid: Control '' accessed from a thread other than the thread it was created on.'
>	System.Windows.Forms.dll!System.Windows.Forms.Control.Handle.get() Line 2502	C#
 	System.Windows.Forms.Primitives.dll!Interop.User32.GetWindowThreadProcessId(IHandle hWnd, out uint lpdwProcessId) Line 17	C#
 	System.Windows.Forms.dll!System.Windows.Forms.Application.ParkingWindow.RemoveReflectChild() Line 81	C#
 	System.Windows.Forms.dll!System.Windows.Forms.Control.ReflectParent.set(System.Windows.Forms.Control value) Line 3137	C#
 	System.Windows.Forms.dll!System.Windows.Forms.Control.UpdateReflectParent(bool findNewParent) Line 11630	C#
 	System.Windows.Forms.dll!System.Windows.Forms.Control.Dispose(bool disposing) Line 5188	C#
 	System.ComponentModel.Primitives.dll!System.ComponentModel.Component.~Component() Line 22	C#
 	[Native to Managed Transition]	
```

### PR DESCRIPTION
**Pull request by M-Lipin**
Fixes #2947 

## Proposed changes

- Changing call to Handle with call to HandleInternal to prevent throwing an error when garbage collection is in progress.

## Customer Impact

- Customers will not see an error in case garbage collector cleans up the references to WinForms controls which are not used anymore.
- The case is applicable on high load environments when running application may require garbage collector to clean up controls to free memory.

## Regression? 

- Yes

## Risk

- Low

### Before

Cross-thread error on garbage collection.

### After

No error.

## Test methodology 

- Manual testing;
- Unit tests (to be implemented);
- UI tests.

## Test environment(s) 

.NET Core SDK (reflecting any global.json):
 Version:   5.0.100-alpha.1.20073.10
 Commit:    29f4d693a9

Runtime Environment:
 OS Name:     Windows
 OS Version:  10.0.18363
 OS Platform: Windows
 RID:         win10-x64
 Base Path:   C:\Program Files\dotnet\sdk\5.0.100-alpha1-05536\

Host (useful for support):
  Version: 5.0.0-alpha.1.20072.3
  Commit:  c3dc1fdfdc

.NET Core SDKs installed:
  3.0.101 [C:\Program Files\dotnet\sdk]
  3.1.200-preview-014883 [C:\Program Files\dotnet\sdk]
  5.0.100-alpha.1.20073.10 [C:\Program Files\dotnet\sdk]
  5.0.100-alpha1-05536 [C:\Program Files\dotnet\sdk]

###### Microsoft Reviewers: [Open in CodeFlow](http://wpcp.azurewebsites.net/CodeFlowProtocolProxy2.php?pullrequest=https://github.com/dotnet/winforms/pull/2951)

### CODE REVIEW

File: src/System.Windows.Forms/src/System/Windows/Forms/Application.ParkingWindow.cs [hunk 1]
```diff
@@ -78,14 +78,14 @@ internal override void RemoveReflectChild()
                 // messagepump is gone and then decide to clean them up.  We should clean
                 // up the parkingwindow in this case and a postmessage won't do it.
 
-                uint id = User32.GetWindowThreadProcessId(this, out _);
+                uint id = User32.GetWindowThreadProcessId(HandleInternal, out _);
```
Comment (weltkante) [inline src/System.Windows.Forms/src/System/Windows/Forms/Application.ParkingWindow.cs]: Issue says you should revert to `new HandleRef(this, HandleInternal)`. While its probably safe during the finalizer run these methods are also used outside a finalizer and passing `HandleInternal` directly without keeping it alive over the call poses a risk of GC calling the finalizer while the call is in progress.

**Comment by M-Lipin** [comment]
Adding tests and test app to reproduce the issue as unit test does not fail due to other thread error (it seems that this is not trivial to catch GC thread error by unit test).
Also added Maui test to reproduce it: Dispose_Parked_ComboBox. But it seems that this also does not fail as expected before the fix.

**Comment by M-Lipin** [comment]
Found out that the issue reproduces (or maybe error catch occurred) only in Debug mode. When running app/test without debugging error does not appear. @RussKie, @JeremyKuhne. @weltkante do you have any thoughts? Simple test app can be found here: src/System.Windows.Forms/tests/IntegrationTests/CrossThreadExceptionTests/CrossThreadExceptionTests/ in the PR branch.

**Comment by weltkante** [comment]
No I don't have any idea about this, timing and JIT generated code may be different (e.g. Debug.Asserts are no longer executed) so something that repros in DEBUG maybe doesn't in RELEASE build. I have no experience how to write unit tests for race conditions or threading problems.

**Comment by JeremyKuhne** [comment]
The GC used to work differently when running debug vs. release binaries, but I don't know if that is still true in Core. In any case it is definitely possible to get things that happen only in debug or release.

**Comment by JeremyKuhne** [comment]
The exception is only thrown when a debugger is attached:

``` C#
         public static bool CheckForIllegalCrossThreadCalls { get; set; } = Debugger.IsAttached;


// In Handle

                if (CheckForIllegalCrossThreadCalls &&
                    !t_inCrossThreadSafeCall &&
                    InvokeRequired)
                {
                    throw new InvalidOperationException(string.Format(SR.IllegalCrossThreadCall,
                                                                     Name));
                }
```

**Comment by RussKie** [comment]
> The exception is only thrown when a debugger is attached:

🤦‍♂ Thank you

@M-Lipin we can set `CheckForIllegalCrossThreadCalls = true` as part of the test, can't we?

**Comment by M-Lipin** [comment]
Added integration test which sets `CheckForIllegalCrossThreadCalls = true` during execution.

File: src/System.Windows.Forms/tests/UnitTests/System/Windows/Forms/Application.ParkingWindowTests.cs [hunk 2]
```diff
@@ -13,6 +13,9 @@ public class ParkingWindowTests
         [WinFormsFact]
         public void ParkingWindow_DoesNotThrowOnGarbageCollecting()
         {
+            bool checkForIllegalCrossThreadCallsDefault = Control.CheckForIllegalCrossThreadCalls;
+            Control.CheckForIllegalCrossThreadCalls = true;
```
Comment (weltkante) [inline src/System.Windows.Forms/tests/UnitTests/System/Windows/Forms/Application.ParkingWindowTests.cs]: Unfortunately this is a static variable, you will be affecting other threads and -worse- if more than one test does this and their execution interleaves you may not reset the variable correctly. I'm not experienced with advanced unit test techniques so I can't suggest a solution. Maybe some kind of fixture is required which ensures that tests requiring `CheckForIllegalCrossThreadCalls` true will execute together but not in parallel with normal tests who need it false.

File: src/System.Windows.Forms/src/System/Windows/Forms/Application.ParkingWindow.cs (hunk 1 above)
Comment (M-Lipin) [reply src/System.Windows.Forms/src/System/Windows/Forms/Application.ParkingWindow.cs]: @RussKie, @weltkante, could you please share your thoughts how to implement the test for finalizer thread?

File: src/System.Windows.Forms/tests/UnitTests/System/Windows/Forms/Application.ParkingWindowTests.cs (hunk 2 above)
Comment (M-Lipin) [reply src/System.Windows.Forms/tests/UnitTests/System/Windows/Forms/Application.ParkingWindowTests.cs]: Thank you for the good point! I've removed this unnecessary unit test as the case is covered by integration test (is executing in a separate process).

Comment (RussKie) [reply src/System.Windows.Forms/tests/UnitTests/System/Windows/Forms/Application.ParkingWindowTests.cs]: There's a remote executor that runs tests as a separate process, we can use that

Comment (M-Lipin) [reply src/System.Windows.Forms/tests/UnitTests/System/Windows/Forms/Application.ParkingWindowTests.cs]: Thank you for the good point! I've implemented integration UI test to cover the current case which is also executed in a separate process: https://github.com/dotnet/winforms/pull/2951/commits/fd1d8851fad388c523f7049351eb4b1da51b0833
Do we still need the unit test here? It seems to me that integration UI test is more appropriate.

**Comment by M-Lipin** [comment]
Unfortunately single unit test cannot be used: unable to get information about GC failure.

**Comment by RussKie** [comment]
Looks like we're the PR builds now. Let's clean up to have only 2 commits:
1. tests that exhibit the failure,
2. code fix that makes the tests pass.

As discussed, I don't think we need a dedicated integration or MAUI tests, unless you see a compelling reason to have those.

**Comment by M-Lipin** [comment]
Thanks! Updated the commit history accordingly.

**Pull request merged**